src/mosquitto_stubs.c

```diff
@@ -381,25 +403,36 @@ void mqtt_callback_dco(struct mosquitto *m, void *obj, int rc) {
   caml_release_runtime_system();
 }
 
+static value extract_qos_list(int qos_count, const int *qos) {
+  CAMLparam0();
+  CAMLlocal2(cli, cons);
+
+  int i;
+
+  cli = Val_emptylist;
+  for (i = qos_count; i > 0; i--) {
+    cons = caml_alloc(2, 0);
+    Store_field(cons, 0, Val_long(qos[i - 1]));
+    Store_field(cons, 1, cli);
+    cli = cons;
+  }
+
+  CAMLreturn(cli);
+}
+
 void mqtt_callback_sub(struct mosquitto *m, void *obj, int mid, int qos_count, const int *qos) {
   struct ocmq *mq;
-  value cli, cons;
-  int i;
+  value cli;
   mq = (struct ocmq*)obj;
+  if (mq->conn == NULL) return;
 
   caml_acquire_runtime_system();
 
   if (NULL == mq->cb[CBSUBSCRIBE]) {
     mq->cb[CBSUBSCRIBE] = caml_named_value(mq->uid[CBSUBSCRIBE]);
   }
   if (NULL != mq->cb[CBSUBSCRIBE]) {
-    cli = Val_emptylist;
-    for (i = qos_count; i > 0; i--) {
-      cons = caml_alloc(2, 0);
-      Store_field(cons, 0, Val_long(qos[i - 1]));
-      Store_field(cons, 1, cli);
-      cli = cons;
-    }
+    cli = extract_qos_list(qos_count, qos);
```

Comment:
complying to the rules of the GC by moving the bad code block to a helper function